Mathlib/Algebra/Module/Injective.lean

```diff
@@ -0,0 +1,467 @@
+          ↑a + ↑b =
+            ↑(ExtensionOfMaxAdjoin.fst i a + ExtensionOfMaxAdjoin.fst i b) +
+              (ExtensionOfMaxAdjoin.snd i a + ExtensionOfMaxAdjoin.snd i b) • y := by
+          rw [ExtensionOfMaxAdjoin.eqn, ExtensionOfMaxAdjoin.eqn, add_smul, Submodule.coe_add]
+          ac_rfl
+        rw [ExtensionOfMaxAdjoin.extensionToFun_wd (y := y) i f h (a + b) _ _ eq1,
+          LinearPMap.map_add, map_add]
+        unfold ExtensionOfMaxAdjoin.extensionToFun
+        abel
+      map_smul' := fun r a => by
+        dsimp
+        have eq1 :
+          r • (a:N) =
+            ↑(r • ExtensionOfMaxAdjoin.fst i a) + (r • ExtensionOfMaxAdjoin.snd i a) • y := by
+          rw [ExtensionOfMaxAdjoin.eqn, smul_add, smul_eq_mul, mul_smul]
+          rfl
+        rw [ExtensionOfMaxAdjoin.extensionToFun_wd i f h (r • a) _ _ eq1, LinearMap.map_smul,
+          LinearPMap.map_smul, ← smul_add]
+        congr }
+  is_extension m := by
+    dsimp
+    simp only [LinearPMap.mk_apply, LinearMap.coe_mk]
+    rw [(extensionOfMax i f).is_extension,
+      ExtensionOfMaxAdjoin.extensionToFun_wd i f h _ ⟨i m, _⟩ 0 _, map_zero, add_zero]
+    simp
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_adjoin Module.Baer.extensionOfMaxAdjoin
+
+set_option synthInstance.etaExperiment true in
+theorem extensionOfMax_le (h : Module.Baer R Q) {y : N} :
+    extensionOfMax i f ≤ extensionOfMaxAdjoin i f h y :=
+  ⟨le_sup_left, fun x x' EQ => by
+    symm
+    change ExtensionOfMaxAdjoin.extensionToFun i f h _ = _
+    rw [ExtensionOfMaxAdjoin.extensionToFun_wd i f h x' x 0 (by simp [EQ]), map_zero,
+      add_zero]⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_le Module.Baer.extensionOfMax_le
+
+theorem extensionOfMax_to_submodule_eq_top (h : Module.Baer R Q) :
+    (extensionOfMax i f).domain = ⊤ := by
+  refine' Submodule.eq_top_iff'.mpr fun y => _
+  dsimp
+  rw [← extensionOfMax_is_max i f _ (extensionOfMax_le i f h), extensionOfMaxAdjoin]
+  rw [Submodule.mem_sup]
+  exact ⟨0, Submodule.zero_mem _, y, Submodule.mem_span_singleton_self _, zero_add _⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of_max_to_submodule_eq_top Module.Baer.extensionOfMax_to_submodule_eq_top
+
+/-- **Baer's criterion** for injective module : a Baer module is an injective module, i.e. if every
+linear map from an ideal can be extended, then the module is injective.-/
+protected theorem injective (h : Module.Baer R Q) : Module.Injective R Q :=
+  { out := fun X Y ins1 ins2 ins3 ins4 i hi f =>
+      haveI : Fact (Function.Injective i) := ⟨hi⟩
+      ⟨{  toFun := fun y =>
+            (extensionOfMax i f).toLinearPMap
+              ⟨y, (extensionOfMax_to_submodule_eq_top i f h).symm ▸ trivial⟩
+          map_add' := fun x y => by
+            rw [← LinearPMap.map_add]
+            congr
+          map_smul' := fun r x => by
+            rw [← LinearPMap.map_smul]
+            dsimp },
```

Comment:
```suggestion
            -- porting note: was `congr`
            dsimp },
```
I also had a `congr` fail on me today in a ported file. Is it worth keeping track of these failures?

Reply:
Probably. I'll record something.

Reply:
Done